engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/UpdateByWindowCumulative.java

```diff
@@ -0,0 +1,274 @@
+        final long keyBefore;
+        try (final RowSet.SearchIterator rIt = context.sourceRowSet.reverseIterator()) {
+            rIt.advance(context.affectedRows.firstRowKey());
+            if (rIt.hasNext()) {
+                keyBefore = rIt.nextLong();
+            } else {
+                keyBefore = NULL_ROW_KEY;
+            }
+        }
+
+        // and preload that data for these operators
+        for (int opIdx : context.dirtyOperatorIndices) {
+            UpdateByCumulativeOperator cumOp = (UpdateByCumulativeOperator) operators[opIdx];
+            if (cumOp.getTimestampColumnName() == null || keyBefore == NULL_ROW_KEY) {
+                // this operator doesn't care about timestamps or we know we are at the beginning of the rowset
+                cumOp.initializeUpdate(context.opContext[opIdx], keyBefore, NULL_LONG);
+            } else {
+                // this operator cares about timestamps, so make sure it is starting from a valid value and
+                // valid timestamp by moving backward until the conditions are met
+                UpdateByCumulativeOperator.Context cumOpContext =
+                        (UpdateByCumulativeOperator.Context) context.opContext[opIdx];
+                long potentialResetTimestamp = context.timestampColumnSource.getLong(keyBefore);
+
+                if (potentialResetTimestamp == NULL_LONG || !cumOpContext.isValueValid(keyBefore)) {
+                    try (final RowSet.SearchIterator rIt = context.sourceRowSet.reverseIterator()) {
+                        if (rIt.advance(keyBefore)) {
+                            while (rIt.hasNext()) {
+                                final long nextKey = rIt.nextLong();
+                                potentialResetTimestamp = context.timestampColumnSource.getLong(nextKey);
+                                if (potentialResetTimestamp != NULL_LONG &&
+                                        cumOpContext.isValueValid(nextKey)) {
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+                // call the specialized version of `intializeUpdate()` for these operators
+                cumOp.initializeUpdate(context.opContext[opIdx], keyBefore, potentialResetTimestamp);
+            }
+        }
+
+        try (final RowSequence.Iterator it = context.affectedRows.getRowSequenceIterator();
+                ChunkSource.GetContext tsGetCtx =
+                        context.timestampColumnSource == null ? null
+                                : context.timestampColumnSource.makeGetContext(context.workingChunkSize)) {
+            while (it.hasMore()) {
+                final RowSequence rs = it.getNextRowSequenceWithLength(context.workingChunkSize);
+                final int size = rs.intSize();
+                Arrays.fill(context.inputSourceChunkPopulated, false);
```

Comment:
Couldn't we just `null` out input source chunks between iterations and then check if the input chunk is `null`?